### PR TITLE
add spmv rule

### PR DESCRIPTION
**Pull request by ZuseZ4**

### CODE REVIEW

**Comment by ZuseZ4** [comment]
I think I got the indexing right, but here I'd really prefer some actual julia tests that run against FD or others.

File: enzyme/Enzyme/Utils.cpp [hunk 1]
```diff
@@ -668,6 +668,197 @@ void callMemcpyStridedLapack(llvm::IRBuilder<> &B, llvm::Module &M,
   B.CreateCall(fn, args, bundles);
 }
 
+void callSPMVDiagUpdate(llvm::IRBuilder<> &B, llvm::Module &M, BlasInfo blas,
+                        IntegerType *IT, llvm::Type *BlasCT,
+                        llvm::Type *BlasFPT, llvm::Type *BlasPT,
+                        llvm::Type *BlasIT, llvm::Type *fpTy,
+                        llvm::ArrayRef<llvm::Value *> args,
+                        const llvm::ArrayRef<llvm::OperandBundleDef> bundles,
+                        bool byRef, bool julia_decl) {
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.cpp:677]:
```suggestion
void callSPMVDiagUpdate(IRBuilder<> &B, Module &M, BlasInfo blas,
                        IntegerType *IT, Type *BlasCT,
                        Type *BlasFPT, Type *BlasPT,
                        Type *BlasIT, Type *fpTy,
                        ArrayRef<Value *> args,
                        ArrayRef<OperandBundleDef> bundles,
                        bool byRef, bool julia_decl) {
```

File: enzyme/Enzyme/Utils.cpp [hunk 2]
```diff
@@ -2265,6 +2456,19 @@ llvm::Value *select_vec_dims(IRBuilder<> &B, llvm::Value *trans,
   return width;
 }
 
+llvm::Value *is_uper(IRBuilder<> &B, llvm::Value *trans, bool byRef) {
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.cpp:2459]:
```suggestion
Value *is_uper(IRBuilder<> &B, Value *trans, bool byRef) {
```

File: enzyme/Enzyme/Utils.h [hunk 3]
```diff
@@ -640,6 +640,14 @@ void callMemcpyStridedLapack(llvm::IRBuilder<> &B, llvm::Module &M,
                              BlasInfo blas, llvm::ArrayRef<llvm::Value *> args,
                              llvm::ArrayRef<llvm::OperandBundleDef> bundles);
 
+void callSPMVDiagUpdate(llvm::IRBuilder<> &B, llvm::Module &M, BlasInfo blas,
+                        llvm::IntegerType *IT, llvm::Type *BlasCT,
+                        llvm::Type *BlasFPT, llvm::Type *BlasPT,
+                        llvm::Type *BlasIT, llvm::Type *fpTy,
+                        llvm::ArrayRef<llvm::Value *> args,
+                        const llvm::ArrayRef<llvm::OperandBundleDef> bundles,
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.h:648]:
```suggestion
                        llvm::ArrayRef<llvm::OperandBundleDef> bundles,
```

File: enzyme/tools/enzyme-tblgen/blas-tblgen.cpp [hunk 4]
```diff
@@ -598,29 +604,24 @@ size_t fwd_call_args(const TGPattern &pattern, size_t actArg,
...
     if (ty == ArgType::len) {
       result.append((Twine("arg_") + name).str());
-    } else if (ty == ArgType::fp) {
+    } else if (ty == ArgType::fp || ty == ArgType::ap ||
+               ty == ArgType::vincData) {
       if (pos == actArg) {
         result.append((Twine("d_") + name).str());
       } else {
         result.append((Twine("arg_") + name).str());
-        // result.append((Twine("fp_") + name).str());
```
Comment (tgymnich) [inline enzyme/tools/enzyme-tblgen/blas-tblgen.cpp:606]: do we still need this?

**Review by wsmoses — APPROVED** [review]

**Comment by tgymnich** [comment]
this PR broke macOS CI

**Pull request merged**